pkg/kubecost/asset.go

```diff
@@ -57,51 +57,69 @@ type Asset interface {
+// to determine if two Assets should be combined. Passing `nil` `aggregateBy` indicates
+// that all available `AssetProperty` keys should be used. Passing empty `aggregateBy` indicates that
+// no key should be used (e.g. to aggregate all assets). Passing one or more `aggregateBy`
+// values will key by only those values.
+// Valid values of `aggregateBy` elements are strings which are an `AssetProperty`, and strings prefixed
+// with `"label:"`.
+func key(a Asset, aggregateBy []string) (string, []error) {
+	var errors []error
 	keys := []string{}
 
-	if props == nil {
-		props = []AssetProperty{
-			AssetProviderProp,
-			AssetAccountProp,
-			AssetProjectProp,
-			AssetCategoryProp,
-			AssetClusterProp,
-			AssetTypeProp,
-			AssetServiceProp,
-			AssetProviderIDProp,
-			AssetNameProp,
+	if aggregateBy == nil {
+		aggregateBy = []string{
+			string(AssetProviderProp),
+			string(AssetAccountProp),
+			string(AssetProjectProp),
+			string(AssetCategoryProp),
+			string(AssetClusterProp),
+			string(AssetTypeProp),
+			string(AssetServiceProp),
+			string(AssetProviderIDProp),
+			string(AssetNameProp),
 		}
 	}
 
-	for _, prop := range props {
+	for _, s := range aggregateBy {
 		switch true {
-		case prop == AssetProviderProp && a.Properties().Provider != "":
+		case s == string(AssetProviderProp) && a.Properties().Provider != "":
 			keys = append(keys, a.Properties().Provider)
-		case prop == AssetAccountProp && a.Properties().Account != "":
+		case s == string(AssetAccountProp) && a.Properties().Account != "":
 			keys = append(keys, a.Properties().Account)
-		case prop == AssetProjectProp && a.Properties().Project != "":
+		case s == string(AssetProjectProp) && a.Properties().Project != "":
 			keys = append(keys, a.Properties().Project)
-		case prop == AssetClusterProp && a.Properties().Cluster != "":
+		case s == string(AssetClusterProp) && a.Properties().Cluster != "":
 			keys = append(keys, a.Properties().Cluster)
-		case prop == AssetCategoryProp && a.Properties().Category != "":
+		case s == string(AssetCategoryProp) && a.Properties().Category != "":
 			keys = append(keys, a.Properties().Category)
-		case prop == AssetTypeProp && a.Type().String() != "":
+		case s == string(AssetTypeProp) && a.Type().String() != "":
 			keys = append(keys, a.Type().String())
-		case prop == AssetServiceProp && a.Properties().Service != "":
+		case s == string(AssetServiceProp) && a.Properties().Service != "":
 			keys = append(keys, a.Properties().Service)
-		case prop == AssetProviderIDProp && a.Properties().ProviderID != "":
+		case s == string(AssetProviderIDProp) && a.Properties().ProviderID != "":
 			keys = append(keys, a.Properties().ProviderID)
-		case prop == AssetNameProp && a.Properties().Name != "":
+		case s == string(AssetNameProp) && a.Properties().Name != "":
 			keys = append(keys, a.Properties().Name)
+		case strings.HasPrefix(s, "label:"):
+			if label := a.Labels()[s[6:]]; label != "" {
```

Comment:
This works in this particular case, but be careful accessing strings by index in Go, as there are some gotchas relating to the fact that you're technically accessing bytes rather than runes. (Read this if you're curious: https://blog.golang.org/strings)

Instead, let's try to use the `strings` package:
https://golang.org/pkg/strings/#TrimPrefix

Reply:
Also, just a question about behavior. If you ask for "label:", then you will get `__unallocated__` for every asset, right? I don't know what the "right" answer is here, but would you say that is correct and would be WAI?

Reply:
I think supplying `?aggregate=label:` returns `__unallocated__` for every asset, unless there's some way to give an asset a label where the key is empty string. I guess maybe it'd be better to just ignore the key in that case? Or send a `400 Empty labels not allowed`-type response.

I'll switch to using `TrimPrefix`. On the behavior, I think it depends on whether we want an API that says "you asked for assets to be grouped on label ' ', and you got it", or "that makes no sense, so it isn't allowed". I'd probably lean toward the latter now that you point it out.

Reply:
I actually like the idea of `400 Bad Request: must provide a label`, just to explicitly avoid the issue altogether. Not a requirement (if users do this and get "unexpected" results, it's kind of on them...) but it's a good idea.